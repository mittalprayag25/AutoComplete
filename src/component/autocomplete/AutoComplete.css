@import "../../constants/variables.css";
@font-face {
    font-family: GillSans;
    src: url(../../assets/fonts/GillSans.ttf);
  }

  @font-face {
    font-family: GillSansBold;
    src: url(../../assets/fonts/GillSansBold.ttf);
  }

.mainContainer{
    width : 40%;
}

@media screen and (max-width: 700px) {
    .mainContainer{
        width : 80%;
    }
}
.searchableContainer{
    width: 100%;
}
.searchInput{
    width : 95%;
    line-height: 35px;
    margin-top: 5px;
    background-image: url(./../../assets/search.png);
    background-size: 18px 18px;
    background-repeat: no-repeat;
    background-position: 10px;
    padding-left: 40px;
    border: none;
    margin-left: 10px;
    font-family: GillSans;
    font-size: 18px;
    flex: 1;
}

.selectContainer {
    width : 100%;
    border: none;
}
.options{
    padding: 15px 20px;
    font-size: 16px;
    font-family: GillSans;
    border: 0.1px dotted var(--color-grey);
}
.selectedOptions{
    padding: 5px 10px;
    border: 1px solid var(--color-options);
    border-radius: 20%;
    background-color: var(--color-options);
    margin: 1px 1px;
    font-size: 12px;
    font-family: GillSans;
    color: var(--color-white);
}
.multiSelectContainer{
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    border: 1px solid var(--color-options);
    flex: 3;
    padding: 5px 0px;
}

.selectedOptionContainer{
    display:flex;
    flex-direction: row;
    overflow:auto;
    margin-left: 10px;
    flex: 2;
    flex-wrap: wrap;
}
.selectedOptionContainer::-webkit-scrollbar {
    display: none;
  }
.cross{
    margin-left: 5px;
    color: var(--color-red);
}

input {
    outline:none;
}

select {
    outline:none;
}