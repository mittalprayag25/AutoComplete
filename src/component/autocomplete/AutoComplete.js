import React, { useState, useEffect } from 'react';
import useComponentVisible from "../../hooks/useComponentVisible";
import './AutoComplete.css';

const AutoComplete = (props) => {
    const [dictionary, setDictionary] = useState([]);
    const { options, onChange, value, onTextChange, multiValues, onRemove } = props;
    const { refSelect, isComponentVisible, setIsComponentVisible } = useComponentVisible(true);

    useEffect(() => {
        setDictionary(options);
    }, [options]);


    const searchText = (value) => {
        setIsComponentVisible(true);
        if (value.length !== 0) {
            const filteredWords = options.filter(data => data.toLowerCase().includes(value.toLowerCase()));
            setDictionary(filteredWords);
            onTextChange(value);
        } else {
            setDictionary(options);
            onTextChange(null);
        }
    }

    const handleChange = (e) => {
        onChange([e.target.value]);
    };

    const enterToSelect = (e) => {
        if (e.keyCode === 13) {
            onChange([e.target.value]);
        }
    };

    const focusToOptions = (e) => {
        if (e.keyCode === 40) {
            e.target.blur();
            document.getElementById("selectContainer").focus();
        }
    };

    return (
        <div className={"mainContainer"}>
            <div className={"searchableContainer"}>
                <div className="multiSelectContainer">
                    {
                        multiValues.length > 0 ?
                            <div className={"selectedOptionContainer"} >
                                {
                                    multiValues.map((item, key) => {
                                        return (
                                            <span data-testid="selectedOptions"
                                                className={"selectedOptions"}>{`${item}`}
                                                <span
                                                    data-testid="removeItem"
                                                    className="cross" onClick={() => onRemove(item)}>x</span>
                                            </span>
                                        )
                                    })
                                }
                            </div> : null
                    }
                    <input
                        className={"searchInput"}
                        type='text'
                        id='searchInput'
                        data-testid="searchInput"
                        placeholder="Search.."
                        onKeyDown={focusToOptions}
                        onKeyUp={(e) => searchText(e.target.value)} />
                </div>
                {dictionary.length > 0 && value ?
                    <select
                        className={"selectContainer"}
                        multiple={true}
                        data-testid="selectContainer"
                        ref={refSelect}
                        id="selectContainer"
                        onKeyUp={enterToSelect}
                        onClick={handleChange}
                        autoFocus>
                        {
                            isComponentVisible ?
                                dictionary.map((item) => {
                                    return (
                                        <option
                                            className={"options"}
                                            value={item}>{item}</option>
                                    )
                                }) : null
                        }
                    </select> : null
                }
            </div>
        </div>
    );
}

export default AutoComplete;