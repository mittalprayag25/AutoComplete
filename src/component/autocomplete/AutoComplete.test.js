import React from 'react';
import { render, fireEvent, act, screen } from '@testing-library/react';
import AutoComplete from './AutoComplete';

test('Renders AutoComplete', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: ['India', 'Singapore', 'China'],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { asFragment } = render(<AutoComplete {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Test onKeyUp in input', () => {
    const setDictionary = jest.fn();
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: ['India', 'Singapore', 'China'],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { getByTestId } = render(<AutoComplete {...props} />);
    act(() => {
        fireEvent.keyUp(getByTestId('searchInput'), { target: { value: 'america' } });
    });
    expect(props.onTextChange).toHaveBeenCalledTimes(1);

    expect(getByTestId('searchInput').value).toBe('america');

    act(() => {
        fireEvent.keyUp(getByTestId('searchInput'), { target: { value: '' } });
    });
    expect(props.onTextChange).toHaveBeenCalledTimes(2);
    expect(getByTestId('searchInput').value).toBe('');
});

test('Test options with no multi values', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { queryByTestId } = render(<AutoComplete {...props} />);

    expect(queryByTestId('selectedOptions')).toBeNull();

});

test('Test selectCOntainer when no options', () => {
    const updatedProps = {
        value: "America",
        options: [],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    };

    const { queryByTestId, getByTestId } = render(<AutoComplete {...updatedProps} />);
    act(() => {
        fireEvent.keyUp(getByTestId('searchInput'), { target: { value: 'america' } });
    });

    expect(queryByTestId('selectContainer')).toBeNull();
});

test('Test Input with keyDOWN code 40', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }

    const { queryByTestId, getByTestId } = render(<AutoComplete {...props} />);
    act(() => {
        fireEvent.keyDown(getByTestId('searchInput'), { target: { value: 'america' }, key: "Enter", keyCode: 40 });
    });

    expect(queryByTestId('selectContainer')).not.toBeNull();
});

test('Test Input with keyDOWN with keydown code not 40', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }

    const { queryByTestId, getByTestId } = render(<AutoComplete {...props} />);
    act(() => {
        fireEvent.keyDown(getByTestId('searchInput'), { target: { value: 'america' }, key: "Enter", keyCode: 40 });
    });

    expect(queryByTestId('selectContainer')).not.toBeNull();
});

test('Test select Container', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { queryByTestId } = render(<AutoComplete {...props} />);

    expect(queryByTestId('selectContainer')).not.toBeNull();
});



test('Test select Container when input text is empty', () => {
    const props = {
        value: "",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { queryByTestId } = render(<AutoComplete {...props} />);

    expect(queryByTestId('selectContainer')).toBeNull();
});

test('Test select Container when option is clicked', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { getByTestId } = render(<AutoComplete {...props} />);
    act(() => {
        fireEvent.click(getByTestId('selectContainer'));
    });
    expect(props.onChange).toHaveBeenCalledTimes(1);

});


test('Test select Container when option is entered by pressing enter key', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: [],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { getByTestId } = render(<AutoComplete {...props} />);
    act(() => {
        fireEvent.keyUp(getByTestId('selectContainer'), { key: "Enter", keyCode: 13 });
    });
    expect(props.onChange).toHaveBeenCalledTimes(1);

});

test('Test Remove Function', () => {
    const props = {
        value: "America",
        options: ['India', 'Singapore', 'China', 'America', 'Australia', 'Kenya', 'South Afirca', 'Iraq', 'Canada'],
        onChange: jest.fn(),
        multiValues: ['India'],
        onRemove: jest.fn(),
        onTextChange: jest.fn(),
    }
    const { getByTestId } = render(<AutoComplete {...props} />);
    act(() => {
        fireEvent.click(getByTestId('removeItem'));
    });
    expect(props.onRemove).toHaveBeenCalledTimes(1);

});
