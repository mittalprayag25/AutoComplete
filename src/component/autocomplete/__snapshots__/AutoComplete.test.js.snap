// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders AutoComplete 1`] = `
<DocumentFragment>
  <div
    class="mainContainer"
  >
    <div
      class="searchableContainer"
    >
      <div
        class="multiSelectContainer"
      >
        <div
          class="selectedOptionContainer"
        >
          <span
            class="selectedOptions"
            data-testid="selectedOptions"
          >
            India
            <span
              class="cross"
              data-testid="removeItem"
            >
              x
            </span>
          </span>
          <span
            class="selectedOptions"
            data-testid="selectedOptions"
          >
            Singapore
            <span
              class="cross"
              data-testid="removeItem"
            >
              x
            </span>
          </span>
          <span
            class="selectedOptions"
            data-testid="selectedOptions"
          >
            China
            <span
              class="cross"
              data-testid="removeItem"
            >
              x
            </span>
          </span>
        </div>
        <input
          class="searchInput"
          data-testid="searchInput"
          id="searchInput"
          placeholder="Search.."
          type="text"
        />
      </div>
      <select
        class="selectContainer"
        data-testid="selectContainer"
        id="selectContainer"
        multiple=""
      >
        <option
          class="options"
          value="India"
        >
          India
        </option>
        <option
          class="options"
          value="Singapore"
        >
          Singapore
        </option>
        <option
          class="options"
          value="China"
        >
          China
        </option>
        <option
          class="options"
          value="America"
        >
          America
        </option>
        <option
          class="options"
          value="Australia"
        >
          Australia
        </option>
        <option
          class="options"
          value="Kenya"
        >
          Kenya
        </option>
        <option
          class="options"
          value="South Afirca"
        >
          South Afirca
        </option>
        <option
          class="options"
          value="Iraq"
        >
          Iraq
        </option>
        <option
          class="options"
          value="Canada"
        >
          Canada
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
